README.md

# fronni
A Python library for quickly calculating & displaying machine learning model performance metrics with confidence intervals.

Data scientists spend a lot of time evaluating the performance of their machine learning models. A common means of doing so is a classification report, such as the one below built-into the [scikit-learn library](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html) (referenced over 440,000 times on Github).


The problem with depending on this is that when you have imbalanced and/or small datasets in your test sample, the volume of data that makes up each point estimate can be very small. As a result, whatever estimates you have for the precision and recall values will likely be unrepresentative of what happens in the real world. Instead, it would be far better to present to the reader a range of values from confidence intervals.

We can easily create confidence intervals for any metric of interest by using the [bootstrap](https://en.wikipedia.org/wiki/Bootstrapping_%28statistics%29#:~:text=This%20technique%20allows%20estimation%20of,sampling%20from%20an%20approximating%20distribution) technique to create hundreds of samples from the original dataset (with replacement), and then throw away the most extreme 5% of values to get a 95% confidence interval, for example.




## Examples

With a pandas dataframe named 'df', you can generate your metrics like this:

    from fronni.classification import classification_report
    report = classification_report(df['label'], df['predicted'], n=1000)
    print(report)

## Requirements

* Python >= 3.6

* numba

* numpy

* scikit-learn

* plotly



## Installing fronni

pip install fronni



## How fronni works

When the datasets are large, running the bootstrap as described above becomes a computationally expensive calculation. Even a relatively small sample of a million rows, when resampled 1,000 times, turns into a billion-row dataset and causes the calculations to take close to a minute. This isn’t a problem when doing these calculations offline, but engineers are often working interactively in a Jupyter notebook and want answers quickly. For this reason we decided to create a few fast implementations of these calculations for common machine learning metrics, such as precision & recall using the [numba library](http://numba.pydata.org/), which provides a speedup of approximately 23X over regular Python parallel-processing code.



## Full documentation

Functions from the classification module:

### classification_report

Generates confidence intervals for precision, recall, & F1 metrics for a binary or multi-class classification model, given arrays of predicted & label values.

| Parameter | Type | Default |
|--|--|--|
| label | Numpy array or Pandas series | None
| predicted | Numpy array or Pandas series | None
| n | integer, number of bootstrap iterations | 1,000
| confidence_level | integer value between 1 & 100 | 95
| as_dict | Boolean, return nested dictionary if True otherwise Pandas dataframe | False
| confidence_level | value between 1 & 100 | 95
| sort_by_sample_size | Boolean, return the Pandas dataframe, sorted in descending order of class sample size | False

### plot_classification_report

Plots precision, recall, & confidence intervals for F1 metrics for a binary or multi-class classification model, given a classification report input.

| Parameter | Type | Default |
|--|--|--|
| report | output from classification_report | None
| save_to_filename | string, path of filename image to save like "image.png" | None

From the regression module:

### regression_report

Generates confidence intervals for RMSE, MAE, and R^2 metrics for a regression model, given arrays of predicted & label values.

| Parameter | Type | Default |
|--|--|--|
| label | Numpy array or Pandas series | None
| predicted | Numpy array or Pandas series | None
| n | integer, number of bootstrap iterations | 1,000
| as_dict | Boolean, return nested dictionary if True otherwise Pandas dataframe | False

See the [CONTRIBUTING](CONTRIBUTING.md) file for how to help out.



## License

fronni is Apache 2.0 licensed, as found in the LICENSE file.


CHANGELOG.md
1.0.0 (January 13, 2021)

### Fronni

* initial release


tests.py
Copyright (c) Facebook, Inc. and its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

import logging
import sys
import unittest

import numpy as np
import pandas as pd
from numba.typed import List
from sklearn.datasets import make_gaussian_quantiles
from sklearn.metrics import classification_report as sklearn_classification_report
from sklearn.metrics import confusion_matrix as sklearn_confusion_matrix
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

from .classification import (
    create_numba_list,
    classification_report,
    metrics_from_confusion_matrix,
    confusion_matrix,
    plot_classification_report,
)
from .regression import regression_report

logger = logging.getLogger(__name__)
handler = logging.StreamHandler(sys.stdout)
handler.setLevel(logging.DEBUG)
logger.addHandler(handler)
logger.setLevel(logging.DEBUG)

RANDOM_SEED = 27


class TestMultiClassClassification:
    @classmethod
    def setup_class(cls):
        """"""
        # Construct dataset
        X1, y1 = make_gaussian_quantiles(
            cov=3.0,
            n_samples=10000,
            n_features=2,
            n_classes=5,
            random_state=RANDOM_SEED,
        )
        X1 = pd.DataFrame(X1)
        cls.label = pd.Series(y1).to_numpy()
        np.random.seed(RANDOM_SEED)
        cls.predicted = pd.Series(y1)[
            np.random.choice(len(y1), size=len(y1), replace=False)
        ].to_numpy()

        cls.labels = create_numba_list(np.unique(cls.label))
        cls.n_labels = len(cls.labels)
        df = pd.DataFrame({"label": cls.label, "predicted": cls.predicted})
        cls.df_np = df.to_numpy()
        cls.matrix = confusion_matrix(
            cls.df_np[:, 0], cls.df_np[:, 1], labels=cls.labels, n_labels=cls.n_labels
        )

    def test_confusion_matrix(self):
        sklearn_matrix = sklearn_confusion_matrix(self.label, self.predicted)
        assert np.array_equal(self.matrix, sklearn_matrix)

    def test_metrics(self):
        """
        put into a identically-formatted dictionaries to be able to use == operator
        """
        report_sklearn = sklearn_classification_report(
            self.label, self.predicted, output_dict=True
        )
        formatted_report = List()
        for key, val in report_sklearn.items():
            if key not in ["accuracy", "macro avg", "weighted avg"]:
                formatted_report.append(val["precision"])
                formatted_report.append(val["recall"])
                formatted_report.append(val["f1-score"])
        report = metrics_from_confusion_matrix(self.matrix, self.labels)
        assert formatted_report == report

    def test_confidence_intervals(self):
        """
        validate that the lower is always <= upper
        and that the point estimate lies within
        """
        np.random.seed(RANDOM_SEED)
        report = classification_report(self.label, self.predicted)
        assert (report["lower"] <= report["upper"]).all()


class TestBinaryClassification:
    @classmethod
    def setup_class(cls):
        """"""
        # Construct dataset
        X1, y1 = make_gaussian_quantiles(
            cov=3.0,
            n_samples=10000,
            n_features=2,
            n_classes=2,
            random_state=RANDOM_SEED,
        )

        cls.label = pd.Series(y1).to_numpy()
        np.random.seed(RANDOM_SEED)
        cls.predicted = pd.Series(y1)[
            np.random.choice(len(y1), size=len(y1), replace=False)
        ].to_numpy()
        cls.labels = create_numba_list(np.unique(cls.label))
        n_labels = len(cls.labels)
        df = pd.DataFrame({"label": cls.label, "predicted": cls.predicted})
        df_np = df.to_numpy()
        cls.matrix = confusion_matrix(
            df_np[:, 0], df_np[:, 1], labels=cls.labels, n_labels=n_labels
        )

    def test_confusion_matrix(self):
        sklearn_matrix = sklearn_confusion_matrix(self.label, self.predicted)
        assert np.array_equal(self.matrix, sklearn_matrix)

    def test_metrics(self):
        """
        put into a identically-formatted dictionaries to be able to use == operator
        """
        report_sklearn = sklearn_classification_report(
            self.label, self.predicted, output_dict=True
        )
        formatted_report = List()
        for key, val in report_sklearn.items():
            if key not in ["accuracy", "macro avg", "weighted avg"]:
                formatted_report.append(val["precision"])
                formatted_report.append(val["recall"])
                formatted_report.append(val["f1-score"])

        report = metrics_from_confusion_matrix(self.matrix, self.labels)
        assert formatted_report == report

    def test_confidence_intervals(self):
        """
        validate that the lower is always <= upper
        and that the point estimate lies within
        """
        np.random.seed(RANDOM_SEED)
        report = classification_report(self.label, self.predicted)
        assert (report["lower"] <= report["upper"]).all()


class TestBootstrap:
    @classmethod
    def setup_class(cls, input_type="integer"):
        """"""
        # Construct dataset
        X1, y1 = make_gaussian_quantiles(
            cov=3.0,
            n_samples=10000,
            n_features=2,
            n_classes=2,
            random_state=RANDOM_SEED,
        )

        if input_type == "string":
            covert_array_to_string = np.vectorize(
                lambda x: "True" if x == 1 else "False"
            )
            y1 = covert_array_to_string(y1)
        cls.label = pd.Series(y1).to_numpy()
        np.random.seed(RANDOM_SEED)
        cls.predicted = pd.Series(y1)[
            np.random.choice(len(y1), size=len(y1), replace=False)
        ].to_numpy()
        # cls.labels = List(np.unique(cls.label))
        cls.df = pd.DataFrame({"label": cls.label, "predicted": cls.predicted})
        cls.df_np = cls.df.to_numpy()

    def test_num_rows(self):
        """
        verify we get df_length samples back
        """
        df_length = self.df.shape[0]
        df_local = self.df_np[
            np.random.choice(df_length, size=df_length, replace=True), :
        ]
        assert df_local.shape[0] == df_length

    def test_randomness(self):
        """
        verify the resampled dataframe is different from the original
        """
        df_length = self.df.shape[0]
        df_local = self.df_np[
            np.random.choice(df_length, size=df_length, replace=True), :
        ]
        assert np.array_equal(self.df, df_local) is False

    def test_confidence_intervals(self):
        """
        validate that the lower is always <= upper
        and that the point estimate lies within
        """
        np.random.seed(RANDOM_SEED)
        report = classification_report(self.label, self.predicted)
        assert (report["lower"] <= report["upper"]).all()


class TestBootstrap_Regression:
    @classmethod
    def setup_class(cls):
        """"""
        # Construct dataset
        mu = np.array([0.0, 0.0])
        num_samples = 10000
        # The desired covariance matrix.
        r = np.array(
            [
                [1.414, 1.414],
                [1.414, 1.414],
            ]
        )
        y = np.random.multivariate_normal(mu, r, size=num_samples)
        cls.y_true = y[:, 0]
        cls.y_pred = y[:, 1]
        cls.df_dict = regression_report(y[:, 0], y[:, 1], as_dict=True)
        cls.df = regression_report(y[:, 0], y[:, 1], as_dict=False)

    def test_r2_score(self):
        """
        verify we get the r2 score
        """
        scikit_r2_score = round(r2_score(self.y_true, self.y_pred), 2)
        numba_r2_score = round(self.df_dict["r2"]["true"], 2)

        assert scikit_r2_score == numba_r2_score

    def test_mae(self):
        """
        verify we get the r2 score
        """
        scikit_mae = round(mean_absolute_error(self.y_true, self.y_pred), 2)
        numba_mae = round(self.df_dict["mae"]["true"], 2)

        assert scikit_mae == numba_mae

    def test_rmse(self):
        """
        verify we get the r2 score
        """
        scikit_rmse = round(np.sqrt(mean_squared_error(self.y_true, self.y_pred)), 2)
        numba_rmse = round(self.df_dict["rmse"]["true"], 2)

        assert scikit_rmse == numba_rmse

    def test_confidence_intervals(self):
        """
        validate that the lower is always <= upper
        and that the point estimate lies within
        """

        assert (self.df["lower"] <= self.df["upper"]).all()


class TestPlots:
    @classmethod
    def setup_class(cls):
        """"""
        # Construct dataset
        X1, y1 = make_gaussian_quantiles(
            cov=3.0,
            n_samples=10000,
            n_features=2,
            n_classes=2,
            random_state=RANDOM_SEED,
        )

        cls.label = pd.Series(y1).to_numpy()
        np.random.seed(RANDOM_SEED)
        cls.predicted = pd.Series(y1)[
            np.random.choice(len(y1), size=len(y1), replace=False)
        ].to_numpy()
        cls.labels = create_numba_list(np.unique(cls.label))
        n_labels = len(cls.labels)
        df = pd.DataFrame({"label": cls.label, "predicted": cls.predicted})
        df_np = df.to_numpy()
        cls.matrix = confusion_matrix(
            df_np[:, 0], df_np[:, 1], labels=cls.labels, n_labels=n_labels
        )

    def test_plot(self):
        """
        verify that we're able to plot the results without errors
        """
        np.random.seed(RANDOM_SEED)
        report = classification_report(self.label, self.predicted)
        assert plot_classification_report(report) == None


class fronni_unittests(unittest.TestCase):
    def test_bootstrap_binary_integer_input(self):
        bootstrap_test_instance = TestBootstrap()
        bootstrap_test_instance.setup_class()
        bootstrap_test_instance.test_num_rows()
        bootstrap_test_instance.test_randomness()
        bootstrap_test_instance.test_confidence_intervals()

    def test_bootstrap_binary_string_input(self):
        bootstrap_test_instance = TestBootstrap()
        bootstrap_test_instance.setup_class(input_type="string")
        bootstrap_test_instance.test_num_rows()
        bootstrap_test_instance.test_randomness()
        bootstrap_test_instance.test_confidence_intervals()

    def test_binary_classification(self):
        binary_classification_test_instance = TestBinaryClassification()
        binary_classification_test_instance.setup_class()
        binary_classification_test_instance.test_confusion_matrix()
        binary_classification_test_instance.test_metrics()
        binary_classification_test_instance.test_confidence_intervals()

    def test_multi_class_classification(self):
        multi_class_classification_test_instance = TestMultiClassClassification()
        multi_class_classification_test_instance.setup_class()
        multi_class_classification_test_instance.test_confusion_matrix()
        multi_class_classification_test_instance.test_metrics()
        multi_class_classification_test_instance.test_confidence_intervals()

    def test_bootstrap_regression(self):
        bootstrap_test_instance = TestBootstrap_Regression()
        bootstrap_test_instance.setup_class()
        bootstrap_test_instance.test_r2_score()
        bootstrap_test_instance.test_mae()
        bootstrap_test_instance.test_rmse()
        bootstrap_test_instance.test_confidence_intervals()

    def test_plotting(self):
        plotting_test_instance = TestPlots()
        plotting_test_instance.setup_class()
        plotting_test_instance.test_plot()


benchmarks.py
Copyright (c) Facebook, Inc. and its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

import logging
import sys
import pandas as pd
from functools import wraps
import time
from collections import defaultdict
import math
from itertools import repeat
import numpy as np
from concurrent.futures import ProcessPoolExecutor
import multiprocessing
from sklearn.metrics import classification_report as sklearn_classification_report
from .classification import classification_report
from sklearn.datasets import make_gaussian_quantiles

logger = logging.getLogger(__name__)
handler = logging.StreamHandler(sys.stdout)
handler.setLevel(logging.DEBUG)
logger.addHandler(handler)
logger.setLevel(logging.DEBUG)

RANDOM_SEED = 27

def timeit(my_func):
    @wraps(my_func)
    def timed(*args, **kw):

        tstart = time.time()
        output = my_func(*args, **kw)
        tend = time.time()

        print('"{}" took {:.3f} ms to execute\n'.format(my_func.__name__, (tend - tstart) * 1000))
        return output
    return timed

class TestPerformance:

    def __init__(self, n_samples=10000, bootstrap_samples=1000):
        self.n_samples = n_samples
        self.bootstrap_samples = bootstrap_samples
        self.start_index = math.floor(bootstrap_samples*0.025)
        self.end_index = math.floor(bootstrap_samples*0.975)
        self.create_dataset()
        self.test_pandas_sklearn_series()
        self.test_pandas_sklearn_parallel()
        self.test_numba_nosklearn_parallel()

    def create_dataset(self):
        X1, y1 = make_gaussian_quantiles(cov=3.,
                                        n_samples=self.n_samples, n_features=2,
                                        n_classes=5, random_state=RANDOM_SEED)
        X1 = pd.DataFrame(X1)
        self.label = pd.Series(y1).to_numpy()
        np.random.seed(RANDOM_SEED)
        self.predicted = pd.Series(y1)[np.random.choice(len(y1), size=len(y1), replace=False)].to_numpy()

        #self.labels = List(np.unique(self.label))
        #self.n_labels = len(self.labels)
        self.df = pd.DataFrame({'label': self.label, 'predicted': self.predicted})

    def interval_calculator(self, values):
        sorted_values = sorted(values)
        return [sorted_values[self.start_index], sorted_values[self.end_index]]

    @timeit
    def test_pandas_sklearn_series(self):
        """
        resample in pandas
        call classification_report n times
        """
        results = defaultdict(lambda: defaultdict(list))
        for _i in range(self.bootstrap_samples):
            resampled = self.df.sample(n = self.n_samples, replace=True)
            report = sklearn_classification_report(resampled['label'], resampled['predicted'], output_dict=True)
            for key, val in report.items():
                if key not in ['accuracy','macro avg','weighted avg']:
                    for metric_name, metric_val in val.items():
                        if metric_name in ['precision','recall','f1']:
                            results[key][metric_name].append(metric_val)
        intervals = defaultdict(lambda: defaultdict(list))
        for class_name, metrics in results.items():
            for metric_name, values in metrics.items():
                results = self.interval_calculator(values)
                intervals[class_name][metric_name] = results

    @staticmethod
    def parallel_reporter(iteration, df, n_samples):
        resampled = df.sample(n = n_samples, replace=True)
        report = sklearn_classification_report(resampled['label'], resampled['predicted'], output_dict=True)
        # report = classification_report(df['label'], df['predicted'], output_dict=True)
        # store in a nested dict by [class_name][metric_name], where the value is a list
        results = {}
        for key, val in report.items():
            if key not in ['accuracy','macro avg','weighted avg']:
                for metric_name, metric_val in val.items():
                    if metric_name in ['precision','recall','f1']:
                        results[key + '_' + metric_name] = metric_val
        return results

    @staticmethod
    def parallel_interval_calculator(key, values, start_index, end_index):
        sorted_values = sorted(values)
        return [key, sorted_values[start_index], sorted_values[end_index]]

    @timeit
    def test_pandas_sklearn_parallel(self):
        with ProcessPoolExecutor(multiprocessing.cpu_count()) as executor: # processes can't share memory so can't write to estimates dict
            results = list(executor.map(self.parallel_reporter, range(self.bootstrap_samples), repeat(self.df), repeat(self.n_samples)))
            all_results = defaultdict(list)
            for result in results:
                # each result is a dict
                for key, val in result.items():
                    all_results[key].append(val)
            results = list(executor.map(self.parallel_interval_calculator, *zip(*all_results.items()), repeat(self.start_index), repeat(self.end_index)))

    def test_numpy_sklearn_series(self):
        pass

    def test_numpy_nosklearn_series(self):
        pass

    def test_numpy_sklearn_parallel(self):
        pass

    def test_numpy_nosklearn_parallel(self):
        pass

    def test_numba_nosklearn_series(self):
        pass

    @timeit
    def cache_numba(self):
        classification_report(self.label, self.predicted, n=2)

    @timeit
    def test_numba_nosklearn_parallel(self):
        report = classification_report(self.label, self.predicted, n=self.bootstrap_samples)
        print(report)

#test = TestPerformance(n_samples=10000, bootstrap_samples=1000)


classification.py
Copyright (c) Facebook, Inc. and its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

#!/usr/bin/env python3
import logging
import math
import math
import sys
from collections import defaultdict
from collections import defaultdict
from itertools import product

import numba
import numpy as np
import pandas as pd
from numba import prange, jit
from numba import types
from numba.typed import List
from sklearn.utils import check_consistent_length
from sklearn.utils.multiclass import type_of_target


estimates = defaultdict(lambda: defaultdict(list))
intervals = defaultdict(lambda: defaultdict(list))

logger = logging.getLogger(__name__)
handler = logging.StreamHandler(sys.stdout)
handler.setLevel(logging.DEBUG)
logger.addHandler(handler)
logger.setLevel(logging.DEBUG)


@jit(nopython=True, cache=True)
def confusion_matrix(y_true, y_pred, labels, n_labels):
    result = np.zeros((n_labels, n_labels), dtype=np.int32)

    for i in range(len(y_true)):
        result[y_true[i]][y_pred[i]] += 1

    return result


@jit(nopython=True, cache=True)
def metrics_from_confusion_matrix(confusion_matrix, classes):
    metrics = List()
    zero_matrix = np.zeros(confusion_matrix.shape, dtype=numba.int32)

    for i in range(len(classes)):
        tp = confusion_matrix[i, i]

        # FN = row i - TP
        fn_mask = np.copy(zero_matrix)
        fn_mask[i, :] = 1
        fn_mask[i, i] = 0
        fn = np.sum(np.multiply(confusion_matrix, fn_mask))

        # FP = column i - TP
        fp_mask = np.copy(zero_matrix)
        fp_mask[:, i] = 1
        fp_mask[i, i] = 0
        fp = np.sum(np.multiply(confusion_matrix, fp_mask))

        # TN = everything - TP - FN - FP
        # tn_mask = 1 - (fn_mask + fp_mask)
        # tn_mask[i, i] = 0  # for TP
        # tn = np.sum(np.multiply(confusion_matrix, tn_mask))

        # TPR = TP/(TP+FN)
        recall = precision = f1_score = 0
        sum_1 = tp + fn
        if sum_1 > 0:
            recall = tp / sum_1

        # FPR = FP/(FP+TN)
        # fpr = fp / (fp + tn)

        sum_2 = tp + fp
        if sum_2 > 0:
            precision = tp / sum_2

        if precision > 0 and recall > 0:
            f1_score = 2 * precision * recall / (precision + recall)

        metrics.append(precision)
        metrics.append(recall)
        metrics.append(f1_score)

    return metrics


@jit(nopython=True, cache=True)
def reporter(df, df_length, iteration, labels, n_labels, all_results):
    # bootstrap samples
    df_local = df[np.random.choice(df_length, size=df_length, replace=True), :]
    matrix = confusion_matrix(
        df_local[:, 0], df_local[:, 1], labels=labels, n_labels=n_labels
    )
    metrics = metrics_from_confusion_matrix(matrix, labels)
    col = 0
    for value in metrics:
        all_results[iteration][col] = value
        col += 1


@jit(nopython=True, cache=True)
def get_actual_metrics(df, df_length, labels, n_labels, true_results):
    # bootstrap samples
    matrix = confusion_matrix(df[:, 0], df[:, 1], labels=labels, n_labels=n_labels)
    metrics = metrics_from_confusion_matrix(matrix, labels)
    col = 0
    for value in metrics:
        true_results[col] = value
        col += 1


@jit(nopython=True, parallel=True, cache=True)
def run(df, df_length, n, labels, n_labels, headers, start_index, end_index):
    """
    fastest way to do this is to just create a numpy matrix of results ie a dataframe
    append every row of results to the dataframe, as each row will have same headers
    just define class headers:
    classA_precision, classA_recall, classB_precision, etc.
    then calculating inervals is just column-by-column sort, can be done in parallel
    """
    # we know ahead of time the shape of the final matrix
    all_results = np.zeros((n, len(headers)), dtype=types.float64)
    true_results = np.zeros(len(headers), dtype=types.float64)
    for i in prange(n):
        reporter(df, df_length, i, labels, n_labels, all_results)
    get_actual_metrics(df, df_length, labels, n_labels, true_results)
    # we can't do list appends or dictionary assignments in parallel safetly in numba (due to thread safety issues)
    # instead, we pre-allocate a matrix of len(headers)
    intervals = np.zeros((len(headers), 3), dtype=types.float64)
    for i in prange(len(headers)):
        column = all_results[:, i]
        column.sort()  # in-place
        intervals[i][0] = column[start_index]
        intervals[i][1] = column[end_index]
        intervals[i][2] = true_results[i]

    return intervals


def create_numba_list(py_list):
    numba_list = List()
    for item in py_list:
        numba_list.append(item)
    return numba_list


def label_counts(labels):
    """
    return a dictioary of label: count pairs

    we're running an old version of numpy, so we can't do:
    np.array(np.unique(labels, return_counts=True)).T
    when it's available, add back & add the numba decorator
    """
    data = pd.value_counts(labels)
    return data.to_dict()


def classification_report(
    label,
    predicted,
    n=1000,
    confidence_level=95,
    as_dict=False,
    sort_by_sample_size=False,
):
    """
    n: number of bootstrap iterations
    confidence_level: 90, 95, etc., between 1 & 100
    as_dict: return a nested dictionary
    sorted: return the Pandas dataframe, sorted in descending order of class sample size
    """
    assert confidence_level > 0 and confidence_level < 100
    assert (type_of_target(label) == "binary") or (
        type_of_target(label) == "multiclass"
    )
    assert (type_of_target(predicted) == "binary") or (
        type_of_target(predicted) == "multiclass"
    )
    check_consistent_length(label, predicted)

    df = pd.DataFrame({"label": label, "predicted": predicted})
    df["id"] = np.arange(len(df))
    df_length = len(df)

    labels = create_numba_list(np.unique(df["label"]))
    n_labels = len(labels)
    mapping_df = pd.DataFrame({"class": labels, "value": range(len(labels))})
    df_label_encoded = df.merge(mapping_df, left_on="label", right_on="class")
    df_predicted_encoded = df.merge(mapping_df, left_on="predicted", right_on="class")
    df_encoded = df_label_encoded.merge(
        df_predicted_encoded, on="id", suffixes=["_label", "_predicted"]
    )
    df_encoded_np = df_encoded[["value_label", "value_predicted"]].to_numpy()

    # we use the caret character as a separator, so we first verify none of labels contain it
    for label in labels:
        assert "^" not in str(
            label
        ), "please remove the '^' character from your labels, it's used as a delimiter in fronni"
    headers = [
        str(item[0]) + "^" + str(item[1])
        for item in product(labels, ["precision", "recall", "f1_score"])
    ]
    delta = (1 - (confidence_level / 100)) / 2
    start_index = math.floor(n * delta)
    end_index = math.floor(n * (1 - delta))

    results = run(
        df_encoded_np,
        df_length,
        n,
        labels,
        n_labels,
        create_numba_list(headers),
        start_index,
        end_index,
    )

    label_count_dict = label_counts(df["label"].to_numpy())
    dict_results = defaultdict(dict)
    for index, row in enumerate(results):
        header_parts = headers[index].split("^")
        class_name = header_parts[0]
        metric_name = header_parts[1]
        dict_results[class_name][metric_name] = {
            "true": row[2],
            "lower": row[0],
            "upper": row[1],
            "sample_size": label_count_dict.get(class_name),
        }
    if sort_by_sample_size:
        dict_results = dict(
            sorted(
                dict_results.items(),
                key=lambda kv: kv[1]["precision"]["sample_size"],
                reverse=True,
            )
        )
    if as_dict:
        return dict_results
    # otherwise, return the results as a pandas dataframe that can be printed out
    formatted_results = []
    for class_name, metrics in dict_results.items():
        for metric_name, metric_values in metrics.items():
            class_metric_dict = {"class": class_name, "metric": metric_name}
            formatted_results.append(
                {**class_metric_dict, **metric_values}  # combine dicts
            )
    df = pd.DataFrame(formatted_results)
    return df


def plot_classification_report(report, save_to_filename=None):
    """
    save_to_filename: "image.png", "image.pdf", etc.
    documentation here: https://plotly.com/python/error-bars/
    """
    import plotly.graph_objects as go

    report = report.iloc[::-1]  # reverse order, in case 'sort_by_sample_size' was true

    precision_df = report.query("metric=='precision'")
    recall_df = report.query("metric=='recall'")
    f1_df = report.query("metric=='f1_score'")
    classes = f1_df["class"].tolist()
    error_width = (f1_df["upper"] - f1_df["lower"]).tolist()

    fig = go.Figure()

    fig.add_trace(
        go.Scatter(
            x=list(f1_df["true"].to_numpy()),
            y=classes,
            marker_color="blue",
            text=list(f1_df["true"].to_numpy()),
            mode="markers",  # no connecting lines
            error_x=dict(
                type="data",  # value of error bar given in data coordinates
                array=error_width,
                color="blue",
                width=8,
                visible=True,
            ),
            name="F1 score",
        )
    )

    fig.add_trace(
        go.Scatter(
            x=list(precision_df["true"].to_numpy()),
            y=classes,
            marker_color="green",
            text=list(precision_df["true"].to_numpy()),
            mode="markers",
            name="precision",
        )
    )

    fig.add_trace(
        go.Scatter(
            x=list(recall_df["true"].to_numpy()),
            y=classes,
            marker_color="red",
            text=list(recall_df["true"].to_numpy()),
            mode="markers",
            name="recall",
        )
    )

    fig.update_xaxes(range=[0, 1])
    fig.update_yaxes(title_text="Class")
    fig.update_yaxes(
        type="category"
    )  # force it to be categorical so we don't have tick marks

    fig.update_layout(
        autosize=False,
        width=1200,
        height=len(classes) * 20,
        margin=dict(l=50, r=50, b=5, t=5, pad=4),
        paper_bgcolor="white",
    )

    fig.show()
    if save_to_filename:  # requires orca or kaleido installed
        fig.write_image(save_to_filename)
